Adding a Stats Module and Multiplicity Calculator

kmdalton opened the pull request:
This branch attempts to address #59 . In this commit I have added a new `rs.stats` module. This contains a function for computing the multiplicity of a set of miller indices under a given a cell and spacegroup. One can cheaply get the completeness from the output of this function. I have also added tests, and two new methods to `rs.utils` in order to support stats. Since this is a fairly substantial change, I'm going to request that @JBGreisman do a thorough code review.

JBGreisman commented on reciprocalspaceship/__init__.py:
```diff
@@ -15,6 +15,9 @@ def getVersionNumber():
 from reciprocalspaceship import algorithms
 
+# Import stats submodule
+from reciprocalspaceship import stats
```
Let's remove this from the top-level API -- as discussed offline, we can add a method to `DataSet` that wraps this call.

JBGreisman commented on reciprocalspaceship/stats/completeness.py:
```diff
@@ -0,0 +1,56 @@
+import pandas as pd
+import reciprocalspaceship as rs
+
+def compute_multiplicity(hobs, cell, spacegroup, anomalous=False, dmin=None):
```
Let's change this to `compute_redundancy()` to avoid confusion with `DataSet.compute_redundancy()`

JBGreisman commented on reciprocalspaceship/stats/completeness.py:
```diff
@@ -0,0 +1,56 @@
+	'K' : hall[:,1],             
+	'L' : hall[:,2],             
+	'Count' : np.zeros(len(hall)),
+    }, cell=cell, spacegroup=spacegroup).groupby(['H', 'K', 'L']).sum()
```
Is this `groupby().sum()` call necessary? Shouldn't all the entries already be unique?

JBGreisman reviewed (approved):
On the whole, I think this looks good. Couple of general comments, not in any particular order:

- While I see that we can compute completeness with this `compute_redundancy()` function, should the file be named `completeness.py` at this point? 
- I think we should make a top-level `rs.stats.compute_redundancy(DataSet)` and perhaps `rs.stats.compute_completeness(DataSet)` that wraps this function and is user-facing. Do you think we should put this "numpy-backed" implementation in a private method? Or in `utils` where the rest of these types of functions live? 

Once we complete this PR, I can put together the above methods that wrap this and operate directly on unmerged `DataSet` objects for redundancy and unmerged or merged `DataSet` objects for completeness.

kmdalton commented:
- I believe I incorporated all your specific changes
 - I renamed the `completeness.py`  to `redundancy.py` which makes more sense as you noted
 
This function is maybe too useful to be totally hidden away as a private method. I am happy to move it to `utils` at your discretion. I think if we do move it to `utils`, it should not return a `DataSet`. Rather, it should return an `n x 3` array of miller indices and an `n` array of ints or floats. 
 
Let me know what you want to do. I'm happy to do whatever. For me the most useful things in this PR are the `generate_*` functions I added. I'm happy as long as those are not contentious.

JBGreisman reviewed (approved):
Looks good to me. My feeling is that we should move the `rs.stats.compute_redundancy()` method into `utils`, and then I can add functions to `stats/` that operate on DataSet objects by calling to that method.

kmdalton commented:
okay -- i moved `redundancy.py` into `utils/stats.py`. I cleaned up the imports and tests.

JBGreisman reviewed (approved):
Looks good -- I think this makes the most sense as you have it now, and can be easily wrapped to make user-facing, convenience methods that take `DataSet` objects.

Feel free to merge.

kmdalton commented:
Okay. I will merge this.

kmdalton commented:
Oops -- accidentally closed this. When I meant to merge. Nothing to see here, please move along...
